### PR TITLE
Introduce SyGuS datatype API

### PR DESCRIPTION
**Pull request by ajreynol**
This introduces a "SyGuS datatype" class which is a wrapper on top of the Expr-layer Datatype class.  This class is responsible for initializing the sygus constructors and sygus-specific information in a datatype. This is a first step towards a future SyGuS API.

Currently, only SyGuS grammar normalization uses this class, although the Parser, the default grammar constructor, and other custom SyGuS algorithms within CVC4 will be updated to use this interface. 

This PR can be merged independently of #3461.  This PR will make it easier to swap out the Expr-level Datatype implementation when the Node-level Datatypes API is ready.

### CODE REVIEW

File: src/expr/sygus_datatype.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+typedef expr::Attribute<SygusAnyConstAttributeId, bool> SygusAnyConstAttribute;
+
+/**
+ * Keeps the necessary information for initializing a sygus datatype, which
+ * includes the operators, names, print callbacks and list of argument types
+ * for each constructor.
+ *
+ * It also maintains a datatype to represent the structure of the type node.
+ *
+ * Notice that is class is only responsible for setting SyGuS-related
```
Comment (HanielB) [inline src/expr/sygus_datatype.h]: is class -> this class

File: src/expr/sygus_datatype.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ * information regarding the datatype. It is still required that the user
```
Comment (HanielB) [inline src/expr/sygus_datatype.h]: that the user -> from the user

File: src/expr/sygus_datatype.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                                   const std::vector<TypeNode>& consTypes,
+                                   bool prepend)
+{
+  if (prepend)
+  {
+    d_ops.push_back(op);
+    d_cons_names.push_back(std::string(name));
+    d_pc.push_back(spc);
+    d_weight.push_back(weight);
+    d_consArgs.push_back(consTypes);
```
Comment (HanielB) [inline src/expr/sygus_datatype.cpp]: this is not prepending :)

File: src/theory/quantifiers/sygus/sygus_grammar_norm.cpp [hunk 4]
```diff
@@ -155,98 +156,10 @@ Node SygusGrammarNorm::TypeObject::eliminatePartialOperators(Node n)
...
-  sygus_norm->d_dt_all.push_back(d_dt);
-  sygus_norm->d_unres_t_all.insert(d_unres_tn.toType());
-  Trace("sygus-grammar-normalize") << "---------------------------------\n";
-}
-
 void SygusGrammarNorm::TransfDrop::buildType(SygusGrammarNorm* sygus_norm,
-                                             TypeObject& to,
+                                             TypeNode orig,
+                                             TypeNode unres,
+                                             SygusDatatype& to,
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/sygus_grammar_norm.cpp]: change the variable name to reflect the new type (this and other occurrences)

Comment (ajreynol) [reply src/theory/quantifiers/sygus/sygus_grammar_norm.cpp]: After my last commit, this no longer applies (grammar norm still operates on TypeObjects).

**Pull request merged**